### DISCUSSION TITLE
How to return Http error from RunTransaction?

### DISCUSSION THREAD

**Discussion by Oggy4th** [pocketbase/pocketbase#2113] [Q&A]
My expected output is only `"error"`

```
func testTransaction(c echo.Context) error {
	err := database.App.Dao().RunInTransaction(func(txDao *daos.Dao) error {

		return c.JSON(500, errors.New("error").Error())
	})

	if err != nil {
		return err
	}

	return c.JSON(http.StatusOK, "ok")
}
```

It gives this output: `"error" "ok"`

I wanna return different errors with different http error codes from RunTransaction.

**Comment by ganigeorgiev** [comment]
You can return an [`ApiError`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.13.4/apis#ApiError), for example:

```go
return apis.NewBadRequestError("Failed to process the request", someRawDataOrError)
```

**Reply by Oggy4th** [reply]
@ganigeorgiev 
 
Thanks for your reply.

Is there any way using `c echo.Context`?

**Reply by ganigeorgiev** [reply]
I don't understand what do you mean. Why do you need the `echo.Context` for this?

If you return an `error` compatible struct (this includes also `ApiError`) from a route handler it will be processed from the central [`HTTPErrorHandler`](https://github.com/pocketbase/pocketbase/blob/master/apis/base.go#L48).

**Reply by ganigeorgiev** [reply]
If you just want to return the transaction error, then based on your example, you can do something like:
```go
err := database.App.Dao().RunInTransaction(func(txDao *daos.Dao) error {
    return errors.New("error")
})

if err != nil {
    return c.JSON(500, err)
}

return c.JSON(http.StatusOK, "ok")
```

**Reply by Oggy4th** [reply]
There are multiple errors with different status code. So I can't distinguish them from one another.

```
func testTransaction(c echo.Context) error {
	err := database.App.Dao().RunInTransaction(func(txDao *daos.Dao) error {

		userRecord, recErr := database.App.Dao().FindAuthRecordByToken(request.Token, database.App.Settings().RecordAuthToken.Secret)
		if recErr != nil {
			if strings.Contains(recErr.Error(), "expired") {
				return c.JSON(http.StatusBadRequest, errors.New("Expired token"))
			} else {
				return c.JSON(http.StatusUnauthorized, errors.New("Unauthorized token"))
			}

		}

		return nil
	})

	if err != nil {
		return err
	}

	return c.JSON(http.StatusOK, "ok")
}
```
In this example when token is expired I would like to send `Bad Request 400` or `Unauthorized 402` on everything else, but the c.JSON error returned from transaction comes `nil`

**Reply by ganigeorgiev** [reply]
You can either write the response to the body and then return the error, aka:

```go
err := errors.New("Expired token")
c.JSON(http.StatusBadRequest, errors.New("Expired token"))
return err
```

OR just return [`ApiError`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.13.4/apis#ApiError) from the transaction (it has a field `Code` that allows you to set a response status code):


```go
// use any of the factory methods NewBadRequestError, NewForbiddenError, etc.
return apis.NewBadRequestError("Some error message...", nil)

// or if you want to set custom status code
return apis.NewApiError(500, "Some error message...", anyExtraDetails)
```

**Reply by Oggy4th** [reply]
I liked first way. So how right way is it? Anything error would affair in the future?

**Reply by ganigeorgiev** [reply]
It is OK, but I'd recommend to use [`ApiError`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.13.4/apis#ApiError):

> ApiError defines the struct for a basic api error response.

Which seems exactly what you want to do.

This way you can also later move your transaction operation easily out of the route handler as there is no `echo.Context` dependency.

**Reply by Oggy4th** [reply]
ApiError gives unnecessary json body. I already have error code but gives again inside json body.

![image](https://user-images.githubusercontent.com/44497354/226570671-f84019f2-1367-427c-8ac9-6121975600f5.png)

**Reply by ganigeorgiev** [reply]
I don't have the full picture of your use case and what you are trying to do, so I'm not sure what to recommend or suggest.

Keep in mind that all PocketBase API errors are standardized and have the above error response format. This is true also when you don't explicitly write a response body and just return a random `error` from a route handler.
